Tagged content list page

fmlharrison opened the pull request:
https://govuk-topic-pages-protot-pr-17.herokuapp.com/

![screen shot 2018-02-14 at 10 10 50](https://user-images.githubusercontent.com/20514501/36198791-79cf246e-116f-11e8-8d1f-0c7ffb55f9ed.png)

This is the PR for the work I did on creating tagged content list pages. These can be navigated to from a topic page after clicking on a "See more" link that each group of tagged content has. 

I was able to use routing constraints to keep make the decided url structure work.

https://trello.com/c/58FxDRQk/59-add-a-see-all-link-that-takes-users-to-a-list-page-of-the-rest-of-the-content-in-each-supergroup

fmlharrison committed:
Add tagged content list controller and change rummager search logic

This is to show a full tagged content items list by format.
The rummager search can filter by a number of field identifiers, so
we are filtering by format to return all the tagged content for a taxon
of a particular format.

Also made a change to the way we are stubbing call to rummager in tests
by using helper methods in the API Adaptors gem.

Add tagged content show view
Add class names using the BEM rules of CSS naming

fmlharrison committed:
Add routing constraints for document types list page

vanitabarrett commented:
Should we hide the 'See More' link if the taxon has 5 or less content_items of a particular type?

Example: Education only has 1 Document Collection, but we still show the link. https://govuk-topic-pages-protot-pr-17.herokuapp.com/topics/education

fmlharrison commented:
@vanitabarrett - I asked the design team about this and they wanted each group to have a 'See More' link regardless of if they have more than 5 tagged content items or not for the time being. I'm sure are the design progresses this will change.

vanitabarrett commented on app/models/rummager_search.rb:
```diff
@@ -22,10 +23,19 @@ def rummager_search
-      count: PAGE_SIZE_TO_GET_EVERYTHING,
-      filter_taxons: [content_id]
-    }
+    if document_type.nil?
```
There's some repetition here as the params are almost identical for each case apart from the `filter_format`. Perhaps this could be simplified, something like:

```ruby
params = {
  start: 0,
  count: PAGE_SIZE_TO_GET_EVERYTHING,
  filter_taxons: [content_id]
}

if document_type
  params.merge!({filter_format: [document_type]})
end

params
```

fmlharrison replied:
Ah nice. That does seem a more succinct way of doing it. I will change it.

vanitabarrett commented on lib/data/formats.yml:
```diff
@@ -0,0 +1,113 @@
+- asylum_support_decision
+- authored_article
+- business_finance_support_scheme
+- calendar-
```
Stray `-`?

fmlharrison replied:
Yep, removed.

vanitabarrett commented on app/views/tagged_content/show.html.erb:
```diff
@@ -0,0 +1,14 @@
+<div class="list-page">
+  <h2 class="list-page__title"><%= @tagged_content.document_type.humanize %> Items - <%= @tagged_content.search_results_count %></h2>
+  <div>
+    <ul class="list-page__content_list">
```
Following BEM, this should be `list-page__content-list` (dash between content and list, rather than underscore)

fmlharrison replied:
Ah yeah! Thanks for seeing that.

vanitabarrett reviewed (changes requested):
Looks great, just a few minor comments!

leenagupte commented on app/models/taxon.rb:
```diff
@@ -23,13 +23,13 @@ def tagged_content_count(content_id)
   end
 
+  def base_path
+    "/#{@base_path}"
```
Why has base_path been made public here? Is it needed in a later commit?

fmlharrison replied:
It is used in the view to get the base path for the 'See More' link to a tagged content list page
https://github.com/alphagov/govuk-topic-pages-prototype/blob/tagged-content-list-page/app/views/topics/show.html.erb#L41

leenagupte commented on spec/controllers/tagged_content_controller_spec.rb:
```diff
@@ -0,0 +1,28 @@
+      @path = "foo/bar"
+      @document_type = "publication"
+      stub_content_store("/#{@path}", @example_schema)
+      stub_rummage_with_doc_type(@example_schema["content_id"], @document_type)
```
You should be able to use test helper `stub_any_rummager_search` from gds-api-adapters as you aren't dependent on the results of the rummager search in this test.

leenagupte replied:
The same is true for stubbing content-store. You should be able to use `content_store_has_item("/#{@path}")`

fmlharrison replied:
The reason we wrote these test helpers were due to the fact that some of the external stub methods weren't working due to some WebMock dependancy problems, but if you have used them on this project then I'm will give it a go.

fmlharrison replied:
Removing them works fine. I have also refactored the code to remove any redundant mocking that can be done with API Adaptors

leenagupte commented on spec/support/helpers/services_request_helpers.rb:
```diff
@@ -10,14 +10,28 @@ def stub_rummager(content_id)
-      count: RummagerSearch::PAGE_SIZE_TO_GET_EVERYTHING,
-      filter_taxons: [content_id]
-    }
+  def params(content_id, document_type = nil)
```
This looks like a duplicate of the `search_params` method in `RummagerSearch`. Is it possible to DRY this up?

fmlharrison replied:
I removed this method as they are not needed in the new refactor of the code.

leenagupte commented on spec/model/rummager_search_spec.rb:
```diff
@@ -3,11 +3,30 @@
+      expect(search_all_content.document_type).to eq(nil)
+    end
+
+    it "should be able initialize with 2 parameter" do
```
I'm not sure if these tests are needed. It feels like they are testing that you can pass optional parameters to the class?

fmlharrison replied:
As discussed, I will remove these tests as it is testing the functionality of the framework, rather than our code. 😄

leenagupte commented on config/application.rb:
```diff
@@ -12,7 +12,10 @@ module GovukTopicPagesPrototype
-    config.eager_load_paths << Rails.root.join('lib')
+    config.eager_load_paths += %W[
+      #{Rails.root.join('lib')}
+      app/controllers/segment_constraints
```
Why did this need to be added? I thought that every folder in `app` was loaded on boot by default?

fmlharrison replied:
If you remove this you get and uninitialised constant error on the line in the routes.rb file where the FormatConstraint class is used.

https://github.com/alphagov/govuk-topic-pages-prototype/blob/tagged-content-list-page/config/routes.rb#L9

leenagupte commented on spec/support/helpers/services_request_helpers.rb:
```diff
@@ -10,14 +10,28 @@ def stub_rummager(content_id)
     allow(Services.rummager).to receive(:search).with(search_params).and_return(rummager_results)
   end
 
+  def stub_rummage_with_doc_type(content_id, document_type)
```
Typo! I think you meant `stub_rummager_with_doc_type`.

fmlharrison replied:
Yeah!! Good spot.

leenagupte reviewed (changes requested):
@fmlharrison The final product works really, well and the use of constraints in the routes is 👍, however, I think there are a few things that can be tidied up with regards to the tests and the commit history.

In the controller tests, you're not actually testing the responses to content-store and rummager, so you can get away with doing the minimum and using the test-helpers in gds-api adapters:

e.g. `content_store_has_item("/#{@path}")` and `stub_any_rummager_search`

This means that you'll no longer need the `stub_rummage_with_doc_type` and `stub_content_store` helper methods.

In the `stub_rummager` helper method, you are always returning the same response, regardless of the content_id passed in, so you could change the signature to not require any parameters.

You can also change how you are stubbing the call to rummager in `stub_rummager` and remove the need to specify the search params.
e.g remove the `.with` from: 
`allow(Services.rummager).to receive(:search).with(search_params).and_return(rummager_results)` 

I think that having the Gds-Api test helper set up in `ServicesRequestHelper` and `ContentSchemaHelpers` is nice as it's not duplicated across the spec files.

We also spotted that `TaggedContentHelpers` also returns a mock rummager response, but I'm happy for this refactored at a later date, independent from this PR.

We can talk in person tomorrow about ways of making the git history a bit easier to follow.

fmlharrison commented:
@leenagupte Thanks for the comments! 
I have made the changes that you had suggested and have been able to clean up some code that wasn't really doing anything and reduce some repetition.

I have changed some commit messages that we hopefully make it easier to follow along with the commit history.

fmlharrison committed:
Use API Adaptor stub methods in taxon model and topics controller tests

As I removed some custom stubbing methods I had to change how some of
the tests were stubbing calls to content store.

Now the tests use the methods in the API Adaptors
gem to stub requests to Content Store.

fmlharrison committed:
Add “See More” link to topic page so User can navigate to list page

leenagupte reviewed (approved):
👍 Thanks for making the changes!
